### ISSUE TITLE
Mark Saved Activations As Donated Buffers to Inductor 

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#129496]
### 🚀 The feature, motivation and pitch

Saved activations that are unaliased should be annotated as donated buffers to inductor. Donated buffers are buffers which will be considered dead after their first use and may be inplace written to. For example:

```
import torch

@torch.compile()
def relu(x):
    return torch.nn.functional.relu(x)

relu(torch.rand([4096, 4096], device="cuda", requires_grad=True)).sum().backward()
```

Gives graphs:

```
[INFO]:TRACED GRAPH
 ===== Forward graph 0 =====
 /home/eellison/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[4096, 4096][4096, 1]cuda:0"):
         # File: /home/eellison/pytorch/work_dir/tmp13.py:5 in relu, code: return torch.nn.functional.relu(x)
        relu: "f32[4096, 4096][4096, 1]cuda:0" = torch.ops.aten.relu.default(primals_1);  primals_1 = None
        le: "b8[4096, 4096][4096, 1]cuda:0" = torch.ops.aten.le.Scalar(relu, 0)
        return [relu, le]
        

[INFO]:TRACED GRAPH
 ===== Backward graph 0 =====
 <eval_with_key>.1 class GraphModule(torch.nn.Module):
    def forward(self, le: "b8[4096, 4096][4096, 1]cuda:0", tangents_1: "f32[4096, 4096][4096, 1]cuda:0"):
         # File: /home/eellison/pytorch/work_dir/tmp13.py:5 in relu, code: return torch.nn.functional.relu(x)
        full_default: "f32[][]cuda:0" = torch.ops.aten.full.default([], 0.0, dtype = torch.float32, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
        where: "f32[4096, 4096][4096, 1]cuda:0" = torch.ops.aten.where.self(le, full_default, tangents_1);  le = full_default = tangents_1 = None
        return [where]
```    

`le` is a donated buffer and we can inplace write to it. There are couple benefits of buffer donation:

-  Memory use. In-place writing to activation saves allocation of output. In addition, knowing which inputs are donated will let us better calculate high water mark of memory in the scheduler.

- Performance. In-place writes can give (small) performance wins. But another context is that with compiled autograd the cudagraphs checking of [memory consistency from run to run](https://github.com/pytorch/pytorch/blob/b1f486aff9f3ae2f0cec557fe2ce5005e1a192bb/torch/_inductor/cudagraph_trees.py#L1603-L1607) adds appreciable overhead. Almost all of the dead buffers (all ? not sure) should be from activations dying in the backward. This is guaranteed but not known to cudagraphs. We can omit checking inputs which died if they are donated buffers.

We probably want to annotate this in aot autograd.

cc @ezyang @anijain2305 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @zou3519 @bdhirsh, @xmfan @BoyuanFeng

**Comment by bdhirsh** [comment]
This feature would be very cool.

Thinking about it some more, there are a few safety conditions I'm a bit worried about though:

(1) Even if we have a saved activation as an input to a backward graph, that doesn't immediately tell us whether or not that saved activation is **also** used by some compute later in the autograd graph.

(2) We also might not be able to know this information statically: you can have a compiled backward subgraph that takes in an (un-aliased) activation, and it's not until we have the full (dynamic) autograd tape at runtime that we know whether the activation inputs to the current backward subgraph are re-used anywhere else in the autograd engine.

Here's my attempt at trying to think of an example of (2):
```
import torch
from torch.testing._internal.two_tensor import TwoTensor



# we will re-use the graph for g across f1 and f2
@torch.compile(backend="aot_eager_decomp_partition")
def g(activation, param2):
    return torch.matmul(activation, param2)

def f1(inp, param1, param2):
    activation = inp + param1
    return g(activation, param2)

def f2(inp, param1, param2):
    activation = torch.exp(inp + param1)
    return g(activation, param2)

inp = torch.ones(4, 4)
param1 = torch.ones(4, 4, requires_grad=True)
param2 = torch.ones(4, 4, requires_grad=True)

out1 = f1(inp, param1, param2)
out2 = f2(inp, param1, param2)
```

Here we are compiling a matmul where the first input is an activation buffer (that we save for backward, and can potentially donate at backward time to inductor). But:

(1) In the first invocation of g's backward (when we call backward() after `f1`),  it is safe to donate the activation's buffer to inductor during backward graph execution

(2) In the second invocation of g's backward (when we call backward() after `f2`), it is not safe: the activation is an input to the compiled backward, but it is also saved and used in the `ExpBackward()` by the autograd engine.

Just to show the graph - running the above prints this fw/bw graph for `g` (`primals_1` is the activation input to g, and we save its transpose for backward).
```
 ===== Forward graph 0 =====
 /home/hirsheybar/local/b/pytorch/torch/fx/_lazy_graph_module.py class GraphModule(torch.nn.Module):
    def forward(self, primals_1: "f32[4, 4][4, 1]cpu", primals_2: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/tmp.py:9 in g, code: return torch.matmul(activation, param2)
        mm: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(primals_2, primals_1)
        permute: "f32[4, 4][1, 4]cpu" = torch.ops.aten.permute.default(primals_2, [1, 0]);  primals_2 = None
        permute_1: "f32[4, 4][1, 4]cpu" = torch.ops.aten.permute.default(primals_1, [1, 0]);  primals_1 = None
        return [mm, permute, permute_1]


INFO: TRACED GRAPH
 ===== Backward graph 0 =====
 <eval_with_key>.1 class GraphModule(torch.nn.Module):
    def forward(self, permute: "f32[4, 4][1, 4]cpu", permute_1: "f32[4, 4][1, 4]cpu", tangents_1: "f32[4, 4][4, 1]cpu"):
         # File: /home/hirsheybar/local/b/pytorch/tmp.py:9 in g, code: return torch.matmul(activation, param2)
        mm_1: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(permute, tangents_1);  permute = None
        mm_2: "f32[4, 4][4, 1]cpu" = torch.ops.aten.mm.default(tangents_1, permute_1);  tangents_1 = permute_1 = None
        return [mm_1, mm_2]
```

So a couple of thoughts/questions (maybe for @eellison / @albanD / @ezyang):

(1) do we agree that the above means that it is not safe to "burn in" decisions at compile time about which inputs to the graph we can mark as donated activations? Elias would know better but I'd imagine it'll make inductor's life a lot harder if the decision about what buffers can be re-used is only know at runtime. I guess we could always make this user-annotated and/or config-driven, although user annotations that can introduce correctness problems feel pretty bad. 

(2) When we are in the middle of running the backward, do we have a way today of querying how many outstanding references there are to a particular saved_for_backward tensor in the autograd graph? (my guess is that a given activation is only "safe to donate" if it only has one remaining reference at the time we are invoking the current backward node that is saving it)

**Comment by bdhirsh** [comment]
Another option: we *do* burn in the decision at compile time, and add a runtime check that errors loudly if we find out at runtime that the buffer was not actually safe to re-use(technically BC-breaking, maybe worth doing if we think this correctness condition is actually very rare)

**Comment by ezyang** [comment]
I don't think @eellison is proposing that (aliases of) inputs/outputs get donated this way. Only intermediates that do not escape. Because the intermediate does not escape, it is impossible for it to get reused by other nodes in the autograd graph (since it does not escape). In your example, the transpose aliases with an input, so it would not be eligible for donation.

There might be some funny business with `backward(retain_graph=True)` though; although we don't really support this with torch.compile, if we did, we would not know if we could donate buffers or not until backwards run time, which is the point in time we know if the saved activations are being destroyed or not.

**Comment by bdhirsh** [comment]
Ok I talked offline with Ed and he pointed out my misunderstanding - in the example that I put above, the saved activation would show up as an input to the **forward** graph, so we would know to properly not mark it as a donated buffer. I misunderstood Elias's initial comment about "saved activations that are unaliased". I guess we want some logic like:

(1) the outputs to the forward graph of AOTAutograd look something like `[*user_outs, *saved_activations]`

(2) of those saved activations, we want to find any that are **intermediates** of the forward (they are not inputs to the forward graph, and do not alias any inputs), which is how we know that they will be truly unused (they cannot escape the current fw/bw graphs), and mark those as donated.

So in terms of what needs to be changed (ignoring compiled autograd for now):

(1) The outputs of the fw graph that we save for backward are defined here: https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/schemas.py#L496. We need to figure out which of them are intermediates not aliased to any fw inputs

(2) The activations become inputs to the backward graph [here](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/runtime_wrappers.py#L1676) (starting at index `len(ctx.symints)`

We need to set some metadata to tell inductor which indices to the backward graph are considered donated. We could change the `bw_compiler()` schemas directly, although if we want to avoid the risk of breaking BC for non-inductor compiler backends, we can instead just directly add a field like `ViewAndMutationMetadata.bw_donated_indices` containing the indices of the inputs that are donated.

(2) For compiled autograd - we need to avoid accidentally donating buffers that are aliased outside of the current compiled region, so compiled autograd might need to propagate along the indices that AOTAutograd computed (I don't think it can assume that every non-parameter grapharg to the bw can be donated? Since these input-to-bw tensors might be aliased by the user).

(3) As Ed pointed out, once we eventually [support double backward](https://github.com/pytorch/pytorch/issues/91469), we might not actually know which buffers safe to re-use until runtime (some of those buffers might need to be saved for double backward). Maybe we should just make the donated buffer logic a config, and assert that it's off for now when `retain_graph=True`?
